core/src/main/java/org/apache/iceberg/PartitionStatsUtil.java

```diff
@@ -0,0 +1,213 @@
+    PartitionData partitionData = coercedPartitionData(entry.file(), table.specs(), partitionType);
+    record.set(
+        Column.PARTITION_DATA.ordinal(), partitionDataToRecord(partitionType, partitionData));
+    record.set(Column.SPEC_ID.ordinal(), entry.file().specId());
+
+    Snapshot snapshot = table.snapshot(entry.snapshotId());
+    if (snapshot != null) {
+      record.set(Column.LAST_UPDATED_SNAPSHOT_ID.ordinal(), snapshot.snapshotId());
+      record.set(Column.LAST_UPDATED_AT.ordinal(), snapshot.timestampMillis());
+    }
+
+    switch (entry.file().content()) {
+      case DATA:
+        record.set(Column.DATA_FILE_COUNT.ordinal(), 1);
+        record.set(Column.DATA_RECORD_COUNT.ordinal(), entry.file().recordCount());
+        record.set(Column.DATA_FILE_SIZE_IN_BYTES.ordinal(), entry.file().fileSizeInBytes());
+        break;
+      case POSITION_DELETES:
+        record.set(Column.POSITION_DELETE_FILE_COUNT.ordinal(), 1);
+        record.set(Column.POSITION_DELETE_RECORD_COUNT.ordinal(), entry.file().recordCount());
+        break;
+      case EQUALITY_DELETES:
+        record.set(Column.EQUALITY_DELETE_FILE_COUNT.ordinal(), 1);
+        record.set(Column.EQUALITY_DELETE_RECORD_COUNT.ordinal(), entry.file().recordCount());
+        break;
+      default:
+        throw new UnsupportedOperationException(
+            "Unsupported file content type: " + entry.file().content());
+    }
+
+    // TODO: optionally compute TOTAL_RECORD_COUNT based on the flag
+    return record;
+  }
+
+  private static PartitionData coercedPartitionData(
+      ContentFile<?> file, Map<Integer, PartitionSpec> specs, Types.StructType partitionType) {
+    // keep the partition data as per the unified spec by coercing
+    StructLike partition =
+        PartitionUtil.coercePartition(partitionType, specs.get(file.specId()), file.partition());
+    PartitionData data = new PartitionData(partitionType);
+    for (int i = 0; i < partitionType.fields().size(); i++) {
+      Object val = partition.get(i, partitionType.fields().get(i).type().typeId().javaClass());
+      if (val != null) {
+        data.set(i, val);
+      }
+    }
+    return data;
+  }
+
+  private static List<String> scanColumns(ManifestContent content) {
+    switch (content) {
+      case DATA:
+        return BaseScan.SCAN_COLUMNS;
+      case DELETES:
+        return BaseScan.DELETE_SCAN_COLUMNS;
+      default:
+        throw new UnsupportedOperationException("Cannot read unknown manifest type: " + content);
+    }
+  }
+
+  private static void incrementLong(Record toUpdate, Record fromRecord, Column column) {
+    toUpdate.set(
+        column.ordinal(),
+        (long) toUpdate.get(column.ordinal()) + (long) fromRecord.get(column.ordinal()));
+  }
+
+  private static void checkAndIncrementLong(Record toUpdate, Record fromRecord, Column column) {
+    if (toUpdate.get(column.ordinal()) != null && fromRecord.get(column.ordinal()) != null) {
```

Comment:
If toUpdate is null but fromRecord is not null shouldn't we set toUpdate to fromRecord?

Reply:
I also think we don't gain a lot by splitting out the check here, do we have a usecase for incrementLong without checking?

Reply:
> I also think we don't gain a lot by splitting out the check here, do we have a usecase for incrementLong without checking?

There are some optional fields and some required fields, I wanted to check null only for optional fields. Hence I had two separate methods to avoid null check for required fields.  But it may not matter much. I will simplify.